### PR TITLE
New design facilitates node selection and adding leaves

### LINKED ISSUE #114: Import function should distinguish between node selection and adding leaves.

### PR DESCRIPTION
**Pull request by Timbozand**
Closes #114 

For each node, there are now 2 dropdown menus. The one on the left is used for node selection. The option "only if related free-text examples are extracted" will work as it does now, i. e. only include the node for entries for which there is a non-empty cell in the CSV. The option "for all entries" will include the node in all of the entries with the value "unspecified". If some mapping is done when the option "for all entries" is selected, the entries that have a non-empty value in the CSV will get that value and the other ones will get "unspecified".

### CODE REVIEW

**Comment by Timbozand** [comment]
[Import function should distinguish between node selection and adding leaves.](https://trello.com/c/opCgCtiF/150-import-function-should-distinguish-between-node-selection-and-adding-leaves)

**Comment by Drullan** [comment]
Looks ok to me.

File: src/js/import.js [hunk 1]
```diff
@@ -389,35 +396,40 @@
   }
 
   function createjsons(lines, CSVHeaders){
-    var selectedFirst;
+    var selectedNodes = $(".import-node-selection").filter(function(i, el) {
+                          return el.value === "everything";
+                        }).parent().parent().add($(".import-select").filter(function(i, el) {
+                          return el.value !== "unspecified";
+                        }).parent().parent().parent());
```
Comment (smeetz) [inline src/js/import.js]:
```js
function byNodeSelection(i, el) { 
    var node = el.querySelector(".import-node-selection"); 
    return node && node.value === "everything";
}
function byImportSelection(i, el) {
    var select = el.querySelector(".import-select");
    return select && select.value !== "unspecified";
}
function createjsons(..) {
    var nodeSelection = $(".import-all-selects-wrapper").filter(function (i, el) {
        return byNodeSelection(i, el) || byImportSelection(i, el)
    })
    ...
}
```

File: src/js/import.js [hunk 2]
```diff
@@ -389,35 +396,40 @@
...
     var jsonObj = {};
     var serpClassification = {};
-    for(var j=0;j<selectedFirst.length;j++){
-      var onlySelectedInputs = $(selectedFirst[j]).parent().parent().
+    for(var j=0;j<selectedNodes.length;j++){
+      var onlySelectedInputs = $(selectedNodes[j]).
           find(".import-select").filter(function(i, el) {return el.value !== "unspecified";}).
           map((i, e) => e.value).toArray();
-      var currentHeader = serp.find(value => value.display === labels[j]);
+
```
Comment (smeetz) [inline src/js/import.js]: Not sure if it's useful to reuse `byImportSelection`, but it works :>
```js
var onlySelectedInputs = selectedNodes[j]
    .filter(byImportSelection)
    .map((i, el) => el.querySelector('.import-select').value)
    .toArray()
```

File: src/js/import.js [hunk 3]
```diff
@@ -455,20 +467,18 @@
     return currentValue;
   }
 
-  function isValueValid(currentValue, checkboxClassName, specifiedInputsLength){
+  function isValueValid(currentValue, currentNode, includeFor){
```
Comment (smeetz) [inline src/js/import.js]:
```js
function isValueValid(...) {
    // something selection strategy 
    if (includeFor === "everything")
        return true;

    // bla bla this case
    if (currentValue === "unspecified" || currentValue[0] === "unspecified")
        return true;

    // something yes have, no buts 
    var classes = currentNode.classList
    var mustHave = classes.has("researchMustHave") || 
                   classes.has("challengeMustHave")
    return classes.has("import-all-selects-wrapper") && mustHave
}

**Pull request merged**